File: B트랙/4주차/참고 코드/median.py
#https://www.acmicpc.net/problem/1655
import sys
from heapq import heappush, heappop
input = sys.stdin.readline
n = int(input().strip())
max_heap = []
min_heap = []
def balance():
    if not max_heap and not min_heap:
        return
    if min_heap and -max_heap[0] > min_heap[0]:
        mini = -heappop(min_heap)
        maxi = -heappop(max_heap)
        heappush(min_heap,maxi)
        heappush(max_heap,mini)
    if len(max_heap) > len(min_heap) + 1:
        heappush(min_heap,-heappop(max_heap))
for i in range(n):
    a = int(input().strip())
    heappush(max_heap,-a)
    balance()
    print(-max_heap[0])

File: B트랙/4주차/참고 코드/median.java
import java.io.*;
import java.util.*;

public class Main
{
    static PriorityQueue<Integer> max_heap = new PriorityQueue<>(Collections.reverseOrder());
    static PriorityQueue<Integer> min_heap = new PriorityQueue<>();
    public static void balance()
    {
        if (max_heap.isEmpty() && min_heap.isEmpty())
        {
            return;
        }
        if (!min_heap.isEmpty() && max_heap.peek() > min_heap.peek())
        {
            int mini = min_heap.poll();
            int maxi = max_heap.poll();
            max_heap.add(mini);
            min_heap.add(maxi);
        }
        if (max_heap.size() > min_heap.size() + 1)
        {
           min_heap.add(max_heap.poll());
        }
    }
    public static void main(String[] args) throws IOException // https://www.acmicpc.net/problem/1655
    {
        BufferedReader scan = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder ans = new StringBuilder();
        int n = Integer.parseInt(scan.readLine().trim());
        for (int i = 0; i < n; ++i)
        {
            int a = Integer.parseInt(scan.readLine().trim());
            max_heap.add(a);
            balance();
            ans.append(max_heap.peek()).append("\n");
        }
        System.out.println(ans);
        scan.close();
    }
}